#410: Add support for  ErrorBoundary

jgoday opened the pull request:
## Pull request description
See #410 for the original issue.
Implement a default IErrorBoundaryLogger for TestContextBase (only for .net6.0).

### PR meta checklist
- [X] Pull request is targeted at `main` branch for code   
  or targeted at `stable` branch for documentation that is live on bunit.dev.
- [X] Pull request is linked to all related issues, if any.
- [X] I have read the _CONTRIBUTING.md_ document.

### Code PR specific checklist
- [X] My code follows the code style of this project and AspNetCore coding guidelines.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly.
- [x] I have updated the appropriate sub section in the _CHANGELOG.md_.
- [X] I have added, updated or removed tests to according to my changes.
  - [X] All tests passed.

dnfadmin commented:
[![CLA assistant check](https://cla.dotnetfoundation.org/pull/badge/not_signed)](https://cla.dotnetfoundation.org/bUnit-dev/bUnit?pullRequest=418) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla.dotnetfoundation.org/bUnit-dev/bUnit?pullRequest=418) before we can accept your contribution.<br/><br/>:x: jgoday [sign now](https://cla.dotnetfoundation.org/bUnit-dev/bUnit?pullRequest=418)<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla.dotnetfoundation.org/check/bUnit-dev/bUnit?pullRequest=418) it.</sub>

egil commented:
> Hi @egil, I have made a PR (#418) but still have some questions (sorry if I'm too clumsy).

No need to be sorry. Im here to guide you.
 
> For making it work we have to change the DotNet6Version to 6.0.0-preview.4.* (in Directory.Build.props).

That's fine. I was going to do that myself.

> 1. Should I create a separate a ErrorBoundaryLogger (with a ILoggerFactory dependency), maybe called DefaultErrorBoundaryLogger or something?

That might actually be a better idea than having a "NullLogger", since people that want to e.g. add logging output to their test output can then just register an ILoggerFactory and get the output from the ErrorBoundary as well. Really good idea.

> 2. Just for curiosity, why the namespace is called 'Bunit.TestDoubles' ?

The word "test doubles" is a general term for something that replaces "real" things during testing, i.e. https://martinfowler.com/bliki/TestDouble.html.

I noticed that `IErrorBoundaryLogger` is in the `Microsoft.AspNetCore.Components.Web` namespace/package, so that actually means bUnits implementaion should be in bunit.web instead. Sorry about not noticing that before.

In that case, you need to register the `IErrorBoundaryLogger` in the `TestServiceProviderExtensions.AddDefaultTestContextServices` method, instead of in `TestContextBase`.

jgoday commented:
Thanks @egil, I have moved the errorboundary folder to bunit.web (as it depends on Microsoft.AspNetCore.Components.Web).
Added a DefaultBoundaryErrorLogger too.
What impl should we register in AddDefaultTestContextServices ? (ErrorBoundaryNullLogger or DefaultErrorBoundaryLogger)

egil commented:
Do ErrorBoundaryNullLogger have any value anymore? 

I do think so, so just remove it and register DefaultErrorBoundaryLogger instead. By default, a NullLoggerFactory is registered, so that means DefaultErrorBoundaryLogger will receive an NullLogger, and then it effectively becomes ErrorBoundaryNullLogger.

jgoday commented on src/bunit.web/TestDoubles/ErrorBoundary/ErrorBoundaryNullLogger.cs:
```diff
@@ -0,0 +1,30 @@
+	/// <summary>
+	/// Dummy implementation of an IErrorBoundaryLogger (needed for ErrorBoundary component).
+	/// </summary>
+	public class ErrorBoundaryNullLogger : IErrorBoundaryLogger
```
Removed in latest commit.

jgoday commented on src/bunit.web/TestDoubles/ErrorBoundary/DefaultErrorBoundaryLogger.cs:
```diff
@@ -0,0 +1,35 @@
+	/// Default implementation of an IErrorBoundaryLogger (needed for ErrorBoundary component).
+    /// It delegates the implementation of LogErrorAsync to an instance created from ILoggerFactory.
+	/// </summary>
+	public class DefaultErrorBoundaryLogger : IErrorBoundaryLogger
```
Now, it logs the error in the same way.

jgoday commented on src/bunit.web/TestDoubles/ErrorBoundary/DefaultErrorBoundaryLogger.cs:
```diff
@@ -0,0 +1,35 @@
+	/// Default implementation of an IErrorBoundaryLogger (needed for ErrorBoundary component).
+    /// It delegates the implementation of LogErrorAsync to an instance created from ILoggerFactory.
+	/// </summary>
+	public class DefaultErrorBoundaryLogger : IErrorBoundaryLogger
```
Changed in the latest commit.

egil commented on src/bunit.web/Extensions/TestServiceProviderExtensions.cs:
```diff
@@ -50,6 +51,10 @@ public static IServiceCollection AddDefaultTestContextServices(this IServiceColl
 			services.AddSingleton<BunitHtmlParser>();
 			services.AddSingleton<IRenderedComponentActivator, RenderedComponentActivator>();
 
+			// IErrorBoundaryLogger
```
```suggestion
```

No need for redundant exceptions.

egil commented on src/bunit.web/TestDoubles/ErrorBoundary/BunitErrorBoundaryLogger.cs:
```diff
@@ -0,0 +1,40 @@
+#if NET6_0_OR_GREATER
```
This file has a mix of tabs and spaces

egil reviewed (changes requested):
Looks like we're almost there. Just a few style changes.

egil commented on tests/bunit.web.tests/TestDoubles/ErrorBoundary/ErrorBoundaryNullLoggerTest.cs:
```diff
@@ -0,0 +1,20 @@
+{
+	public class ErrorBoundaryNullLoggerTest : TestContext
+	{
+		[Fact(DisplayName = "TestContextBase should provide a default IErrorBoundaryLogger")]
```
Sorry, now that I think of it, then this test belongs in the TestContextBaseTest class.

jgoday replied:
Would it be okay if I make 'TestContextBaseTest' a parial class ? (To allow use it with both net5 and net6 tests, as it's defined now by TestContextBaseTest.net5.cs file)

egil replied:
Absolutely, make it partial. Although, it should probably be `TestContextTest`, since this is something that will be assoicated with it, and not really `TestContextBase`. Sorry for going back and forth on this. Think we have it now :)

jgoday replied:
Ok, I just have changed it.

egil commented on CHANGELOG.md:
```diff
@@ -9,6 +9,7 @@ All notable changes to **bUnit** will be documented in this file. The project ad
 ### Added
 
 List of added functionality in this release.
+- Added `BunitErrorBoundaryLogger` impl of Microsoft.AspNetCore.Components.Web.IErrorBoundaryLogger` (needed for Blazor ErrorBoundary component, from net6.0).
```
```suggestion
- Added `BunitErrorBoundaryLogger` impl of Microsoft.AspNetCore.Components.Web.IErrorBoundaryLogger` (needed for Blazor ErrorBoundary component, from net6.0). By [@jgoday](https://github.com/jgoday).
```

Remember to credit yourself :)

egil commented on tests/bunit.core.tests/TestContextTest.net5.cs:
```diff
@@ -8,7 +8,7 @@
 namespace Bunit
 {
-	public partial class TestContextBaseTest : TestContext
+	public partial class TestContextTest : TestContext
```
Uhm no, the TestContextBaseTest class should stay the same. The tests in there belongs with it.

The test "TestContext should provide a default IErrorBoundaryLogger" do however belong in a `TestContextTest` class, since `IErrorBoundaryLogger` only works with bunit.web, where TestContext exists.

jgoday replied:
sorry @egil, I had misunderstood it. Should be changed now (and moved the errorBoundary test to bunit.web)

egil replied:
It's fault for not clear enough. I'll take a look later tonight.

egil commented on tests/bunit.core.tests/TestContextTest.net6.cs:
```diff
@@ -7,7 +7,7 @@
 namespace Bunit.TestDoubles.ErrorBoundary
 {
-	public partial class TestContextBaseTest : TestContext
+	public partial class TestContextTest : TestContext
```
This belongs in bunit.web.

egil commented on tests/bunit.web.tests/TestContextTest.net6.cs:
```diff
@@ -0,0 +1,20 @@
+{
+	public partial class TestContextTest : TestContext
+	{
+		[Fact(DisplayName = "TestContextBase should provide a default IErrorBoundaryLogger")]
```
```suggestion
		[Fact(DisplayName = "TestContext should provide a default IErrorBoundaryLogger")]
```

egil commented:
Hey @jgoday

It looks like we are almost here. Thanks for your swift reaction to the PR feedback, well done, and apologiens for not being clear in all cases.

There do seem to be [some issues with preview 4 and running the tests](https://github.com/bUnit-dev/bUnit/pull/418/checks), at least here on GitHub, which I will have to investigate before merging. If you have any idea what the problem is, do let me know. Otherwise ill take a crack at it ASAP, but it might not before the weekend, as I have a couple of conference presentations I have to prepare.

jgoday commented:
@egil Thank you so much all your guide/advices/corrections.

Failing tests with 6 preview4 seems to be related to [#53297](https://github.com/dotnet/runtime/issues/53297),
It is happening to me too with preview4 and .net5 (5.0.21) in opensuse (only when multitargeting, if I run the tests with -f net6.0 all works correctly).

egil commented:
> Failing tests with 6 preview4 seems to be related to [#53297](https://github.com/dotnet/runtime/issues/53297),
> It is happening to me too with preview4 and .net5 (5.0.21) in opensuse (only when multitargeting, if I run the tests with -f net6.0 all works correctly).

OK, so a (temporary) solution might be to change the verification.yml workflow to call `dotnet test` three times, once for each targeted framework.

Are you up for making that change?

jgoday commented:
> > Failing tests with 6 preview4 seems to be related to [#53297](https://github.com/dotnet/runtime/issues/53297),
> > It is happening to me too with preview4 and .net5 (5.0.21) in opensuse (only when multitargeting, if I run the tests with -f net6.0 all works correctly).
> 
> OK, so a (temporary) solution might be to change the verification.yml workflow to call `dotnet test` three times, once for each targeted framework.
> 
> Are you up for making that change?

If I understand it correctly would be enough to change verification.yml like this

```
 - name: 🧪 Run unit tests (net3.1)
        run: |
          dotnet test ./tests/bunit.core.tests/bunit.core.tests.csproj -c release --logger GitHubActions -f net3.0
          dotnet test ./tests/bunit.web.tests/bunit.web.tests.csproj -c release --logger GitHubActions -f net3.0
          dotnet test ./tests/bunit.web.testcomponents.tests/bunit.web.testcomponents.tests.csproj -c release --logger GitHubActions  -f net3.0

 - name: 🧪 Run unit tests (net5.0)
        run: |
          dotnet test ./tests/bunit.core.tests/bunit.core.tests.csproj -c release --logger GitHubActions -f net5.0
          dotnet test ./tests/bunit.web.tests/bunit.web.tests.csproj -c release --logger GitHubActions -fnet5.0
          dotnet test ./tests/bunit.web.testcomponents.tests/bunit.web.testcomponents.tests.csproj -c release --logger GitHubActions -f net5.0

 - name: 🧪 Run unit tests (net6.0)
        run: |
          dotnet test ./tests/bunit.core.tests/bunit.core.tests.csproj -c release --logger GitHubActions -f net6.0
          dotnet test ./tests/bunit.web.tests/bunit.web.tests.csproj -c release --logger GitHubActions -f net6.0
          dotnet test ./tests/bunit.web.testcomponents.tests/bunit.web.testcomponents.tests.csproj -c release --logger GitHubActions -f net6.0
```

jgoday commented:
> > > Failing tests with 6 preview4 seems to be related to [#53297](https://github.com/dotnet/runtime/issues/53297),
> > > It is happening to me too with preview4 and .net5 (5.0.21) in opensuse (only when multitargeting, if I run the tests with -f net6.0 all works correctly).
> > 
> > 
> > OK, so a (temporary) solution might be to change the verification.yml workflow to call `dotnet test` three times, once for each targeted framework.
> > Are you up for making that change?
> 
> If I understand it correctly would be enough to change verification.yml like this
> 
> ```
>  - name: 🧪 Run unit tests (net3.1)
>         run: |
>           dotnet test ./tests/bunit.core.tests/bunit.core.tests.csproj -c release --logger GitHubActions -f net3.0
>           dotnet test ./tests/bunit.web.tests/bunit.web.tests.csproj -c release --logger GitHubActions -f net3.0
>           dotnet test ./tests/bunit.web.testcomponents.tests/bunit.web.testcomponents.tests.csproj -c release --logger GitHubActions  -f net3.0
> 
>  - name: 🧪 Run unit tests (net5.0)
>         run: |
>           dotnet test ./tests/bunit.core.tests/bunit.core.tests.csproj -c release --logger GitHubActions -f net5.0
>           dotnet test ./tests/bunit.web.tests/bunit.web.tests.csproj -c release --logger GitHubActions -fnet5.0
>           dotnet test ./tests/bunit.web.testcomponents.tests/bunit.web.testcomponents.tests.csproj -c release --logger GitHubActions -f net5.0
> 
>  - name: 🧪 Run unit tests (net6.0)
>         run: |
>           dotnet test ./tests/bunit.core.tests/bunit.core.tests.csproj -c release --logger GitHubActions -f net6.0
>           dotnet test ./tests/bunit.web.tests/bunit.web.tests.csproj -c release --logger GitHubActions -f net6.0
>           dotnet test ./tests/bunit.web.testcomponents.tests/bunit.web.testcomponents.tests.csproj -c release --logger GitHubActions -f net6.0
> ```

Maybe it is necessary to triplicate the whole 'verify-bunit' job, to avoid dotnet 6 and 5 conflicts in the same setup.

Should I try it in this PR ?

egil commented:
> If I understand it correctly would be enough to change verification.yml like this.

Exactly what I was thinking.

> Maybe it is necessary to triplicate the whole 'verify-bunit' job, to avoid dotnet 6 and 5 conflicts in the same setup.

I hope that is not needed. If it is, then I prefer to just add another "level" to the workflow matrix, and have the framework version be a variable from that, which we use in the test and in the setup-dotnet step.

> Should I try it in this PR ?

Yes please.

egil commented:
Hmm, no dice it seems. Think net3.0 needs to be netcoreapp3.1, but you can check that locally by running the command yourself.

Otherwise, it does seem like we get into trouble with all the frameworks installed.

jgoday commented:
I have changed it to include dotnet versions in strategy.matrix, so each test runs in a completely separated dotnet version.
I think it could work this way.

egil commented:
> I have changed it to include dotnet versions in strategy.matrix, so each test runs in a completely separated dotnet version.
> I think it could work this way.

Thanks for giving this a shoot @jgoday, I think we need to just and see what comes out of the issue you referenced, instead of continue trying to get workarounds working.

egil commented:
Looks like we need to undo the work flow changes and apply the hack mentioned here: https://github.com/dotnet/sdk/issues/18148#issuecomment-860957180

jgoday commented:
Applied https://github.com/dotnet/sdk/issues/18148#issuecomment-860957180 and tested locally.
Reverted verification.yml workflow too.

Need to specify RemoveLoggingAnalyzer individually in bunit.testassets.csproj (to run it before RazorCompileComponentDeclaration because being a Microsoft.NET.Sdk.Razor project)

egil commented:
Awesome, thank you for all the hard work.
